app/models/moab_storage_root.rb

```diff
@@ -4,11 +4,25 @@
 # Metadata about a Moab storage root (a POSIX file system which contains Moab objects).
 class MoabStorageRoot < ApplicationRecord
   has_many :complete_moabs, dependent: :restrict_with_exception
+  has_many :preserved_objects, through: :complete_moabs
   has_and_belongs_to_many :preservation_policies
 
   validates :name, presence: true, uniqueness: true
   validates :storage_location, presence: true
 
+  scope :preserved_objects, lambda {
+    joins(complete_moabs: [:preserved_object])
+  }
+
+  # not a scope because it doesn't return a chainable relation.
+  # this could be a little expensive: use it judiciously.
+  # this completed in seconds on prod for root16, which had the most druids at the time of this writing (314162).  so it's not that bad.
+  def druid_list_sorted
+    preserved_objects
+      .pluck(:druid) # `#pluck` instead of e.g. `#select` keeps us from instantiating a whole AR obj when we just want this string
+      .sort # the list could be large (potentially > 500K; prefer load on app server to load on DB server)
```

Comment:
Why not sort in db? That's what dbs are for.

Reply:
Also, this might be using a cursor for the results. However, you're obviating the value of it by forcing all results into memory.

Reply:
> Why not sort in db? That's what dbs are for

in my testing on prod w/ the largest storage root (4, iirc, just under 350K druids), this returned in seconds (felt like under 5, but i didn't time it).  since it felt like this might be used during normal production activity, i thought it'd be safer to put the load on the app server than on the DB server, though i think the load is negligible either way, so i don't have strong feelings about this.  but my thinking was better to keep the DB as responsive as possible, and to keep queries running for as little time as possible, for deadlock prevention.

and i like DBs more for FK constraints and other schema enforcement and such than for sorting, but i like 'em for sorting too 😛 

> Also, this might be using a cursor for the results. However, you're obviating the value of it by forcing all results into memory.

query and in-app sort ran super quickly, so i don't think this is really an issue here at present, but again, happy to page through instead if you have a strong preference for that.  didn't seem worth the complexity based on my testing, but that was the first approach i considered.

i can see worries about this scaling as storage roots become larger, no preference as to whether we consider that now or kick the can down the road.

could instead return an enumerator here and page through in the CSV writing loop.

once nice thing about the pluck is that no AR objects get instantiated, since we don't need those here, just druid strings.

also, to be totally fair, i tested the query on prod, but not the actual CSV writing (because i could test the query on prod w/o deploying code changes).  doubt the CSV writing would make much of a difference, but i could def be wrong there.

Reply:
> Also, this might be using a cursor for the results

i got curious about this and did a bit of digging.  none of the docs seem to say it directly, but i found some circumstantial evidence indicating that `find_each` doesn't use a cursor, and that in general ActiveRecord doesn't do any automatic cursor magic.  this makes sense -- `find_each` docs are pretty up front about its primary-key-based-paging, non-orderable, multiple-query-running batching strategy.  i also got the impression that postgres will not use a cursor unless you go through the machinations of setting one up (i.e., it didn't seem to me like PG would decide to use a cursor to execute a query via a plain `select` statement).

and in general i was rusty with cursors, so i looked to see how you'd use one in rails-land, and i found:  https://github.com/afair/postgresql_cursor

seems reasonably well-maintained.  i didn't find a ton in the way of (blog, stack exchange, forum) posts about PG cursor usage (or cursors in general) for rails, but i did find a couple people recommending that gem.  it seems reasonably well written/maintained/tested, other than not showing test coverage.

so, here's an alternate implementation that uses that cursor gem:  https://github.com/sul-dlss/preservation_catalog/pull/1347

as a bonus compared to `find_each`, that branch's implementation uses a method from the postgresql_cursor gem that returns each row as a hash containing only the selected columns, instead of a fully instantiated active record object.